include/deal.II/grid/grid_tools.h

```diff
@@ -1727,6 +1727,9 @@ namespace GridTools
                      + cell->face(face)->vertex(2) + cell->face(face)->vertex(3)) / 4.0;
               }
       }
+
+    // Make sure FEValues notices that the mesh has changed
+    triangulation.signals.post_refinement();
```

Comment:
Oh, indeed, crazy. Do we need to call `pre_refinement` at the top of the function?